spring-integration-zeromq/src/main/java/org/springframework/integration/zeromq/ZeroMqProxy.java

```diff
@@ -0,0 +1,281 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.zeromq;
+
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.zeromq.SocketType;
+import org.zeromq.ZContext;
+import org.zeromq.ZMQ;
+
+import org.springframework.beans.factory.BeanNameAware;
+import org.springframework.beans.factory.InitializingBean;
+import org.springframework.context.SmartLifecycle;
+import org.springframework.lang.Nullable;
+import org.springframework.scheduling.concurrent.CustomizableThreadFactory;
+import org.springframework.util.Assert;
+
+/**
+ * This class encapsulates a logic to configure and manage ZeroMQ proxy.
+ * It binds frontend and backend sockets over TCP on all the available network interfaces
+ * with provided ports or randomly selected otherwise.
+ * <p>
+ * The {@link ZeroMqProxy.Type} dictates which pair of ZeroMQ sockets to bind with this proxy
+ * to implement any possible patterns for ZeroMQ intermediary. Defaults to @link {@link ZeroMqProxy.Type#PULL_PUSH}.
+ * <p>
+ * The control socket is exposed as a {@link SocketType#PAIR} with an inter-thread transport
+ * on tne {@code "inproc://" + beanName + ".control"} address - can be obtained via {@link #getControlAddress()}.
+ * Should be used with the same application from {@link SocketType#PAIR} socket to send
+ * {@link zmq.ZMQ#PROXY_TERMINATE}, {@link zmq.ZMQ#PROXY_PAUSE} and/or {@link zmq.ZMQ#PROXY_RESUME} commands.
+ * <p>
+ * If proxy cannot be started for some reason, the error message is logged respectively and this component is
+ * left in the non-starting state.
+ *
+ * @author Artem Bilan
+ *
+ * @since 5.4
+ *
+ * @see ZMQ#proxy(ZMQ.Socket, ZMQ.Socket, ZMQ.Socket)
+ */
+public class ZeroMqProxy implements InitializingBean, SmartLifecycle, BeanNameAware {
```

Comment:
I thought about that, but probably it is harmless since `Executors.newSingleThreadExecutor()` is a `FinalizableDelegatedExecutorService`. 
Although you might mean to still clean up the resource we don't hold any more...